library/lua/datetime.lua

```diff
@@ -0,0 +1,237 @@
+local _ENV = mkmodule('datetime')
+
+--[[
+TODO: investigate applicability for adv mode.
+if advmode then TICKS_PER_DAY = 86400 ? or only for cur_year_tick?
+advmod_TU / 72 = ticks
+--]]
+
+-- are these locals better in DwarfCalendar?
+local DAYS_PER_MONTH = 28
+local DAYS_PER_YEAR = 336
+local MONTHS_PER_YEAR = 12
+local TICKS_PER_DAY = 1200
+local TICKS_PER_MONTH = TICKS_PER_DAY * DAYS_PER_MONTH
+local TICKS_PER_YEAR = TICKS_PER_MONTH * MONTHS_PER_YEAR
+
+local MONTHS = {
+ 'Granite',
+ 'Slate',
+ 'Felsite',
+ 'Hematite',
+ 'Malachite',
+ 'Galena',
+ 'Limestone',
+ 'Sandstone',
+ 'Timber',
+ 'Moonstone',
+ 'Opal',
+ 'Obsidian',
+}
+
+DwarfCalendar = defclass(DwarfCalendar)
+
+DwarfCalendar.ATTRS{
+    year=0,
+    year_tick=0,
+}
+
+function DwarfCalendar:init()
+    self:normalize()
+end
+
+function DwarfCalendar:addTicks(ticks)
+    self.year_tick = self.year_tick + ticks
+    self:normalize()
+end
+
+function DwarfCalendar:addDays(days)
+    self.year_tick = self.year_tick + self.daysToTicks(days)
+    self:normalize()
+end
+
+function DwarfCalendar:addMonths(months)
+    self.year_tick = self.year_tick + self.monthsToTicks(months)
+    self:normalize()
+end
+
+function DwarfCalendar:addYears(years)
+    self.year = self.year + years
+end
+
+-- returns an integer pair: (year), (year tick count)
+function DwarfCalendar:getYears()
+    return self.year, self.year_tick
+end
+
+-- returns days since beginning of a year, starting from zero
+function DwarfCalendar:ticksToDays()
+    return self.year_tick // TICKS_PER_DAY
+end
+
+-- returns days since the beginning of a month, starting from zero
+function DwarfCalendar:ticksToDayOfMonth()
+    return self:ticksToDays() % DAYS_PER_MONTH
+end
+
+-- returns months since the beginning of a year, starting from zero
+function DwarfCalendar:ticksToMonths()
+    return self.year_tick // TICKS_PER_MONTH
+end
+
+-- returns ticks since the beginning of a day
+function DwarfCalendar:getDayTicks()
+    return self.year_tick % TICKS_PER_DAY
+end
+
+-- returns ticks since the beginning of a month
+function DwarfCalendar:getMonthTicks()
+    return self.year_tick % TICKS_PER_MONTH
+end
+
+function DwarfCalendar:normalize()
+    if (self.year_tick > TICKS_PER_YEAR) then
+        self.year = self.year + (self.year_tick // TICKS_PER_YEAR)
+        self.year_tick = self.year_tick % TICKS_PER_YEAR
+    elseif (self.year_tick < 0) then
+        -- going backwards in time, subtract year by at least one.
+        self.year = self.year - math.max(1, math.abs(self.year_tick) // TICKS_PER_YEAR)
+        self.year_tick = self.year_tick % TICKS_PER_YEAR
```

Comment:
won't self.year_tick still be negative here?

Reply:
also do remember that lua's mechanics for `%` with respect to negative values may be unintuitive, due to the fact that lua rounds toward negative infinity instead of toward zero

Reply:
I should have commented that part. self.year_tick will be positive. This behavior is documented in Lua's manual 3.4.1.  

Lua's % operator should be equivalent to:
function m(a, b)
  return a - (b * (a // b))
end

If I understand this correctly, the difference between modulo operator in Lua and C-lang+friends is the direction of quotient rounding via the use of the integer floor op where it effectively rounds toward minus infinity (i.e. -1.123 becomes -2). 

C-lang+friends round toward zero (-1.123 becomes -1). Consequently, in this direction, it ensures the remainder's sign agrees with the dividend, as it should. I believe Lua's math.fmod() rounds toward zero -- so not the same as %.

Incidentally, Python, Ruby, and Perl also use this variant of modulo.  I think it's safe to depend on this behavior going forward.

Reply:
For example:
max year tick = 403200

if year = 2 and year_tick is -1, the result is:
year: 1
year_tick: 403199